adaptive/utils.py

```diff
@@ -67,3 +68,14 @@ def decorator(method):
         return functools.wraps(other)(method)
 
     return decorator
+
+
+class _RequireAttrsABCMeta(abc.ABCMeta):
+    required_attributes = []
+
+    def __call__(self, *args, **kwargs):
+        obj = super().__call__(*args, **kwargs)
+        for name in obj.required_attributes:
+            if not hasattr(obj, name):
+                raise ValueError(f"Required attribute {name} not set in __init__.")
```

Comment:
We should also check that the type is good.

We should use class attribute annotations to specify this:

```
class A:
    x: int
    y: float
```